srml/system/rpc/src/lib.rs

```diff
@@ -73,46 +85,135 @@ where
 	P: txpool::ChainApi + Sync + Send + 'static,
 	Block: traits::Block,
 	AccountId: Clone + std::fmt::Display + Codec,
-	Index: Clone + std::fmt::Display + Codec + traits::SimpleArithmetic,
+	Index: Clone + std::fmt::Display + Codec + Send + traits::SimpleArithmetic + 'static,
 {
-	fn nonce(&self, account: AccountId) -> Result<Index> {
-		let api = self.client.runtime_api();
-		let best = self.client.info().best_hash;
-		let at = BlockId::hash(best);
+	fn nonce(&self, account: AccountId) -> FutureResult<Index> {
+		let get_nonce = || {
+			let api = self.client.runtime_api();
+			let best = self.client.info().best_hash;
+			let at = BlockId::hash(best);
+
+			let nonce = api.account_nonce(&at, account.clone()).map_err(|e| Error {
+				code: ErrorCode::ServerError(RUNTIME_ERROR),
+				message: "Unable to query nonce.".into(),
+				data: Some(format!("{:?}", e).into()),
+			})?;
+
+			Ok(adjust_nonce(&*self.pool, account, nonce))
+		};
+
+		Box::new(result(get_nonce()))
+	}
+}
 
-		let nonce = api.account_nonce(&at, account.clone()).map_err(|e| Error {
+/// An implementation of System-specific RPC methods on light client.
+pub struct LightSystem<P: txpool::ChainApi, C, F, Block> {
+	client: Arc<C>,
+	remote_blockchain: Arc<dyn RemoteBlockchain<Block>>,
+	fetcher: Arc<F>,
+	pool: Arc<Pool<P>>,
+}
+
+impl<P: txpool::ChainApi, C, F, Block> LightSystem<P, C, F, Block> {
+	/// Create new `LightSystem`.
+	pub fn new(
+		client: Arc<C>,
+		remote_blockchain: Arc<dyn RemoteBlockchain<Block>>,
+		fetcher: Arc<F>,
+		pool: Arc<Pool<P>>,
+	) -> Self {
+		LightSystem {
+			client,
+			remote_blockchain,
+			fetcher,
+			pool,
+		}
+	}
+}
+
+impl<P, C, F, Block, AccountId, Index> SystemApi<AccountId, Index> for LightSystem<P, C, F, Block>
+where
+	P: txpool::ChainApi + Sync + Send + 'static,
+	C: HeaderBackend<Block>,
+	C: Send + Sync + 'static,
+	F: Fetcher<Block> + 'static,
+	Block: traits::Block,
+	AccountId: Clone + std::fmt::Display + Codec + Send + 'static,
+	Index: Clone + std::fmt::Display + Codec + Send + traits::SimpleArithmetic + 'static,
+{
+	fn nonce(&self, account: AccountId) -> FutureResult<Index> {
+		let best_hash = self.client.info().best_hash;
+		let best_id = BlockId::hash(best_hash);
+		let future_best_header = future_header(&*self.remote_blockchain, &*self.fetcher, best_id);
+		let fetcher = self.fetcher.clone();
+		let call_data = account.encode();
+		let future_best_header = future_best_header
+			.and_then(move |maybe_best_header| ready(
+				match maybe_best_header {
+					Some(best_header) => Ok(best_header),
+					None => Err(ClientError::UnknownBlock(format!("{}", best_hash))),
+				}
+			));
+		let future_nonce = future_best_header.and_then(move |best_header|
+			fetcher.remote_call(RemoteCallRequest {
+				block: best_hash,
+				header: best_header,
+				method: "AccountNonceApi_account_nonce".into(),
```

Comment:
Hmm, that seems very fragile, can't we store that on the `trait` level as consts to avoid harcoding strings like this in the code?

Reply:
We should at least add a test that this is the same as the runtime API.

Reply:
I've made a separate issue for that - #3875